### PR TITLE
Optimize ResourceContainer/AssemblyLoadEventHandler methods, remove allocs

### PR DESCRIPTION
**Pull request by h3xds1nz**
## Description

Optimizes two rather important startup methods in `ResourceContainer` class, the handler for runtime assembly load events - `OnAssemblyLoadEventHandler` and `GetResourceManagerWrapper`. This saves several milliseconds upon startup.

- I retracted from using `StringComparer.OrdinalIgnoreCase` as it would be slower by ~60ns in the final solution.
- For parsing the assembly name from safe input (`RuntimeAssembly`), I've re-used the #9739 and added a new method for `Version`/`PublicKeyToken`
- I've originally used `ArrayPool<char>` for the unlikely event of needing an array (256+ chars) but seems redundant to initialize it this early, and we shall not really need to allocate an array anyways in 99% cases.
- Using `AlternativeLookup`, this now becomes fully allocation-free given that usually only 1 entry is in the dictionary.
- It is safe to assume that `ToLowerInvariant` produces same-length output, the own function's sanity check and several other dotnet runtime libraries (plus tons of user code) depend on this behavior of per-character folding.
- Some of the comments (e.g. about GACed assemblies) could have been removed, but I've kept them in.
- We seal the class for additional perf benefit as it is derived.

Do note that these benchmarks measure throughput and allocations (accurate), the savings are in milliseconds this early.
Allocations are dependent on presence of public key / assembly name length, this just illustrates general scenario.

### Single throughput of AssemblyLoadHandler

| Method    | assembly             | Mean [ns] | Error [ns] | StdDev [ns] | Code Size [B] | Gen0   | Allocated [B] |
|---------- |--------------------- |----------:|-----------:|------------:|--------------:|-------:|--------------:|
| Original  | Syste(...)7798e [89] |  636.3 ns |    6.57 ns |     6.14 ns |       5,880 B | 0.0896 |        1504 B |
| PR__EDIT | Syste(...)7798e [89] |  121.6 ns |    0.44 ns |     0.39 ns |       5,052 B |      - |             - |

### Standard app startup (28 libraries loaded)

| Method    | assembly             | Mean [ns]   | Error [ns] | StdDev [ns] | Code Size [B] | Gen0   | Allocated [B] |
|---------- |--------------------- |------------:|-----------:|------------:|--------------:|-------:|--------------:|
| Original  | Syste(...)7798e [89] | 17,579.6 ns |  151.20 ns |   141.43 ns |       8,030 B | 2.6245 |       44352 B |
| PR__EDIT | Syste(...)7798e [89] | 3,395.5 ns |    5.43 ns |     4.81 ns |       3,166 B |      - |             - |

## Customer Impact

Improved startup performance, decreased startup allocations.

## Regression

No.

## Testing

Local build, running few sample apps using resources and verifying functionality.

## Risk

Low, I believe I didn't achieve any change of the behavior, even unintentionally.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/wpf/pull/9822)

### CODE REVIEW

**Review by edwardneal — COMMENTED** [review]
Looking good! Does the benchmark handle the startup of an app with something bound to an embedded image resource? There's a call to Assembly.Load in the call stack of ResourceContainer.GetPart (GetPart -> GetPartCore -> GetResourceManagerWrapper -> BaseUriHelper.GetLoadedAssembly -> Assembly.Load) and it feels like this is going to be the largest performance killer in the worst-case scenario; I'm curious to see whether it's triggered under normal circumstances.

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs [hunk 1]
```diff
@@ -243,123 +184,111 @@ protected override PackagePart GetPartCore(Uri uri)
...
-                string assemblyName = assemblyInfo.Name.ToLowerInvariant();
-                string assemblyKey = string.Empty;
-                string key = assemblyName;
+                Span<char> key = totalLength <= 256 ? stackalloc char[totalLength] : new char[totalLength];
+                assemblyName.ToLowerInvariant(key);
 
                 // Check if this newly loaded assembly is in the cache. If so, update the cache.
                 // If it is not in cache, do not do anything. It will be added on demand.
-                // The key could be Name, Name + Version, Name + PublicKeyToken, or Name + Version + PublicKeyToken.  
+                // The key could be Name; Name + Version; Name + PublicKeyToken; or Name + Version + PublicKeyToken.
```
Comment (miloush) [inline src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:207]: (Makes me wonder which language uses semicolons this way, comma was grammatically fine.)

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs [hunk 2]
```diff
@@ -243,123 +184,111 @@ protected override PackagePart GetPartCore(Uri uri)
...
+
+                // First time; add this to the dictionary and create the wrapper if its not the application entry assembly
+                if (!s_registeredResourceManagersLookup.TryGetValue(key.Slice(0, totalLength), out rmwResult))
                 {
-                    Assembly assembly;
-
-                    assembly = BaseUriHelper.GetLoadedAssembly(assemblyName, assemblyVersion, assemblyKey);
-
+                    Assembly assembly = BaseUriHelper.GetLoadedAssembly(assemblyName, assemblyVersion, assemblyToken);
                     if (assembly.Equals(Application.ResourceAssembly))
```
[the remark this answers is not in the document]
Comment (miloush) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:277]: Seems this is run only once and then stored in the cache of resource managers.

File: src/Microsoft.DotNet.Wpf/src/Shared/MS/Internal/ReflectionUtils.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+using System;
+
+namespace MS.Internal
+{
+    /// <summary>
+    /// Provides utilities for working with reflection efficiently.
+    /// </summary>
+    internal static class ReflectionUtils
+    {
+#if !NETFX
```
Comment (miloush) [inline src/Microsoft.DotNet.Wpf/src/Shared/MS/Internal/ReflectionUtils.cs:19]: Is this needed for PBT or something that compiles for .NET Framework? I would love to keep the codebase compilable for .NET FW but I think that ship has already sailed, so if there is no need then you might as well just keep the .NET Core path only.

Comment (h3xds1nz) [reply src/Microsoft.DotNet.Wpf/src/Shared/MS/Internal/ReflectionUtils.cs:19]: Yeah, I wish I didn't need to but it was the most concise thing to do. Trust me, I'd remove any `netfx` reference if I could, haha.

#9739 needs a dummy fallback for PBT

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (h3xds1nz) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:277]: Speedscope, is that what the cool kids use now? Haha.

Anyways as @miloush says, this is usually only run once, you shall not make it past the lookup, I originally had string there for the key but changed it in the last commits as you rarely need to add an entry.

The method itself shall be optimized but I felt it doesn't need to be done in the scope of this PR, however any modification has to wait anyways. I currently have 60 open PRs and I'm waiting for them to be resolved so I can make more changes now because I'm effectively blocking myself in a lot of places to not overlap changes.

A bit frustrating but gotta live with the reality.

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs (hunk 1 above)
Comment (h3xds1nz) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:207]: I'll revert this, shouldn't have been changed. My language might as well be the same as yours, you never know.

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (singhashish-wpf) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:277]: > I currently have 60 open PRs and I'm waiting for them to be resolved so I can make more changes now because I'm effectively blocking myself in a lot of places to not overlap changes.
> 
> A bit frustrating but gotta live with the reality.

Please post the PRs which are blocking (causing overlap), We will try to prioritize those. The next CTP has 5 of your PRs lined up. Agree with the sentiment that this is a slow pace and frustrating, however the test passes take up time, and with the .NET 9 releases and other deliverables, the max we can get is 8 PRs a month.

[the remark this answers is not in the document]
Comment (miloush) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:277]: @singhashish-wpf would it be possible to run the tests on a batch of PRs together? Or perhaps @h3xds1nz could do a bigger PR with all the picked commits?

[the remark this answers is not in the document]
Comment (singhashish-wpf) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:277]: We are going to batch in 10. The problem is we want the haystack to be smaller in case of test failures, otherwise we could have taken all of them together.

[the remark this answers is not in the document]
Comment (h3xds1nz) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:277]: @singhashish-wpf Thank you, that would be very lovely if we could prioritize. I understand the amount of tests run is rather large for WPF, it takes time (was it even a few days?) and you're trying to minimize the number so it is easier to pinpoint in case of encountering case failures. Nonetheless, I'm very glad you're taking the time for the discussion here.

As per the priorities, this would help a bunch:

1) #9363 
- This will unblock me on rather a big PR in terms of allocations with most converters (but small in code changes).
2) #9676 / #9697 and #9683
-  This will unblock optimizations of Key/MouseGesture converters as a result.
3) #9220
- I keep constantly hitting on code with changes I made here.
4) #9711 
- Here I'd mostly like to know ahead if we're gonna go that direction, because then I could add the dead code removal that will end up as a result directly into the PR. The save of that is like 12ms during startup.
5) #9481
- This would be nice to have faster for further optimizations on the call-chains I had in mind.
6) #9292
- This PR is from Thomas but would help a great deal to be squeezed in.
7) #9521
- This directly makes a modification in the `BaseUriHelper` for example. The goal is to get rid of the CAS helpers as it should have been done few years ago before your team has stepped into WPF.

I'd also have one question - since I've been mostly writing tests for all the changes, should I be adding them in? My issue is that currently it is hard to not duplicate some tests from wpf-test (any movement on the intended move of the wpf-test repo into main repo as per the discussion in #9401?) and few assemblies don't even have Test projects, the assemblies would require `InternalsVisibleTo` or at least use of UnsafeAccessors/reflection ain't the best idea, I know some DRTs still depend on it though.

[the remark this answers is not in the document]
Comment (singhashish-wpf) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:280]: Thanks @h3xds1nz for the list. I will get these added in the next CTP. cc:// @harshit7962 

If you are writing unit tests, we still have some in the System.Xaml area and others, you can add them in this repo only. If you want to leverage the DRT setup, you can do it on wpf-test repo. We are looking for better ways to have a common place and an integrated testing infra. Tests at both the places are validated before PR merge, so as of now submitting at any place is fine. The gist of the discussion is hinting towards movement of the tests from wpf-test repo, which is going to take up time/effort, also we ideally don't want to keep following the old testing infra as that has execution challenges and learning curve for contributors.

[the remark this answers is not in the document]
Comment (h3xds1nz) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs:280]: @singhashish-wpf Thanks a bunch. Yeah not keen on the DRT setup, thanks for the guidance regarding unit tests.

Happy to hear that things are moving in some direction :)

**Comment by JeremyKuhne** [comment]
@adamsitnik wanted to point this out to you as it is doing some assembly name parsing in a hot path for WPF.

**Review by himgoyalmicro — APPROVED** [review]

**Comment by himgoyalmicro** [comment]
Thank you @h3xds1nz for your continuous contributions 😄

**Comment by h3xds1nz** [comment]
@himgoyalmicro Thank you, happy this one is in <3
